bin/update-tables.py

```diff
@@ -1,332 +1,446 @@
+
+This should be executed through tox,
+
+    $ tox -e update
+
+If data files were previously downloaded, but will refresh by last-modified
+check using HEAD request from unicode.org URLs, unless --no-check-last-modified
+is used:
+
+    $ tox -e update -- --check-last-modified
 
 https://github.com/jquast/wcwidth
 """
-from __future__ import print_function
+from __future__ import annotations
 
 # std imports
 import os
 import re
-import glob
-import codecs
+import sys
 import string
-import urllib
+import logging
 import datetime
-import collections
+import functools
 import unicodedata
+from pathlib import Path
+from dataclasses import field, fields, dataclass
+
+from typing import Any, Mapping, Iterable, Iterator, Sequence, Container, Collection
+from typing_extensions import Self
 
-try:
-    # py2
-    from urllib2 import urlopen
-except ImportError:
-    # py3
-    from urllib.request import urlopen
+# 3rd party
+import jinja2
+import requests
+import urllib3.util
+import dateutil.parser
 
-URL_UNICODE_DERIVED_AGE = 'http://www.unicode.org/Public/UCD/latest/ucd/DerivedAge.txt'
+URL_UNICODE_DERIVED_AGE = 'https://www.unicode.org/Public/UCD/latest/ucd/DerivedAge.txt'
+URL_EASTASIAN_WIDTH = 'https://www.unicode.org/Public/{version}/ucd/EastAsianWidth.txt'
+URL_DERIVED_CATEGORY = 'https://www.unicode.org/Public/{version}/ucd/extracted/DerivedGeneralCategory.txt'
 EXCLUDE_VERSIONS = ['2.0.0', '2.1.2', '3.0.0', '3.1.0', '3.2.0', '4.0.0']
-PATH_UP = os.path.relpath(
-    os.path.join(
-        os.path.dirname(__file__),
-        os.path.pardir))
-PATH_DOCS = os.path.join(PATH_UP, 'docs')
+
+PATH_UP = os.path.relpath(os.path.join(os.path.dirname(__file__), os.path.pardir))
 PATH_DATA = os.path.join(PATH_UP, 'data')
-PATH_CODE = os.path.join(PATH_UP, 'wcwidth')
-FILE_RST = os.path.join(PATH_DOCS, 'unicode_version.rst')
-FILE_PATCH_FROM = "release files:"
-FILE_PATCH_TO = "======="
-
-
-# use chr() for py3.x,
-# unichr() for py2.x
-try:
-    _ = unichr(0)
-except NameError as err:
-    if err.args[0] == "name 'unichr' is not defined":
-        # pylint: disable=C0103,W0622
-        #         Invalid constant name "unichr" (col 8)
-        #         Redefining built-in 'unichr' (col 8)
-        unichr = chr
-    else:
-        raise
+# "wcwidth/bin/update-tables.py", even on Windows
+# not really a path, if the git repo isn't named "wcwidth"
+THIS_FILEPATH = ('wcwidth/' +
+                 Path(__file__).resolve().relative_to(Path(PATH_UP).resolve()).as_posix())
 
+JINJA_ENV = jinja2.Environment(
+    loader=jinja2.FileSystemLoader(os.path.join(PATH_UP, 'code_templates')),
+    keep_trailing_newline=True)
+UTC_NOW = datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S UTC")
 
-TableDef = collections.namedtuple('table', ['version', 'date', 'values'])
+CONNECT_TIMEOUT = int(os.environ.get('CONNECT_TIMEOUT', '10'))
+FETCH_BLOCKSIZE = int(os.environ.get('FETCH_BLOCKSIZE', '4096'))
+MAX_RETRIES = int(os.environ.get('MAX_RETRIES', '10'))
```

Comment:
10 retries may be too many here, the tenth retry will wait for `0.1 * (2 ** (10 - 1)) == 51.2` seconds. 6 or 7 retries are enough.